### ISSUE TITLE
cmd/vet: crash of go vet involving cgo

### ISSUE THREAD

**Issue by rfielding** [golang/go#18389]
### What version of Go are you using (`go version`)?
go version go1.7.4 darwin/amd64

```
Robs-MBP:object-drive-server rfielding$ go env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/rfielding/gocode"
GORACE=""
GOROOT="/Users/rfielding/go1.7.4"
GOTOOLDIR="/Users/rfielding/go1.7.4/pkg/tool/darwin_amd64"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/nk/ftxlzcmx5yg0v6bnkm940zlh0000gn/T/go-build124531480=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
```

`go vet` crashes with an infinite recursion in cgo.go:

```
go vet ./...
...

runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x2d130d, 0xe)
	/usr/local/go/src/runtime/panic.go:566 +0x95
runtime.newstack()
	/usr/local/go/src/runtime/stack.go:1061 +0x416
runtime.morestack()
	/usr/local/go/src/runtime/asm_amd64.s:366 +0x7f

goroutine 1 [running]:
main.typeOKForCgoCall(0x3f0b20, 0xc4204b8560, 0xc4203ee660)
	/usr/local/go/src/cmd/vet/cgo.go:111 fp=0xc4407002b8 sp=0xc4407002b0
main.typeOKForCgoCall(0x3f0ae0, 0xc42049c360, 0xc4204b8590)
	/usr/local/go/src/cmd/vet/cgo.go:124 +0x104 fp=0xc4407002f8 sp=0xc4407002b8
main.typeOKForCgoCall(0x3f0b20, 0xc4204b8590, 0xc4203ee601)
	/usr/local/go/src/cmd/vet/cgo.go:119 +0x160 fp=0xc440700338 sp=0xc4407002f8
...
main.typeOKForCgoCall(0x3f0ae0, 0xc42049c360, 0xc4204b8590)
	/usr/local/go/src/cmd/vet/cgo.go:124 +0x104 fp=0xc440701b78 sp=0xc440701b38
...additional frames elided...
exit status 2
vendor/gopkg.in/yaml.v2/decode.go:123: unreachable code
vendor/gopkg.in/yaml.v2/emitterc.go:669: unreachable code
vendor/gopkg.in/yaml.v2/parserc.go:169: unreachable code
exit status 1
```

This code uses the spacemonkey/openssl library.  The error happens during vet in cgo.go:

```
	// typeOKForCgoCall returns true if the type of arg is OK to pass to a
   109	// C function using cgo. This is not true for Go types with embedded
   110	// pointers.
   111	func typeOKForCgoCall(t types.Type) bool {
   112		if t == nil {
   113			return true
   114		}
   115		switch t := t.Underlying().(type) {
   116		case *types.Chan, *types.Map, *types.Signature, *types.Slice:
   117			return false
   118		case *types.Pointer:
   119			return typeOKForCgoCall(t.Elem())
   120		case *types.Array:
   121			return typeOKForCgoCall(t.Elem())
   122		case *types.Struct:
   123			for i := 0; i < t.NumFields(); i++ {
   124				if !typeOKForCgoCall(t.Field(i).Type()) {
   125					return false
   126				}
   127			}
   128		}
   129		return true
   130	}
   131	
```

This seems plausible if `t` has a `t.Field(i).Type()` reaches the same type.  In my case, it reaches the same type with period 2.

```
main.typeOKForCgoCall(0x3f0b20, 0xc420493f30, 0xc42046d001)
	/usr/local/go/src/cmd/vet/cgo.go:119 +0x160 fp=0xc440701a38 sp=0xc4407019f8
main.typeOKForCgoCall(0x3f0ae0, 0xc42049cc60, 0xc420493f30)
	/usr/local/go/src/cmd/vet/cgo.go:124 +0x104 fp=0xc440701a78 sp=0xc440701a38
main.typeOKForCgoCall(0x3f0b20, 0xc420493f30, 0xc42046d001)
	/usr/local/go/src/cmd/vet/cgo.go:119 +0x160 fp=0xc440701ab8 sp=0xc440701a78
```

**Comment by bradfitz** [comment]
With Go 1.8 coming out soon, we're no longer making changes to the Go 1.7 codebase. Can you check whether this bug is present in Go 1.8?

```
$ go get golang.org/x/build/version/go1.8beta2
$ go1.8beta2 download
$ go1.8beta2 vet ./...
```

**Comment by rfielding** [comment]
yes.  i am getting 1.8 to build from source.  this looks like an easy change.

**Comment by rfielding** [comment]
I got the same problem with 1.8.  It is alternating between `openssl.Certificate` and `*openssl.Certificate` in the infinite recursion on the type being passed in.

**Comment by rfielding** [comment]
and I tried it with the `go1.8beta2` binary as well:
```
Robs-MacBook-Pro:object-drive-server rfielding$ go1.8beta2 vet ./...
vendor/github.com/c9s/goprocinfo/linux/netstat.go:98: struct field TCPOFODrop repeats json tag "tcp_ofo_drop" also at vendor/github.com/c9s/goprocinfo/linux/netstat.go:97
exit status 1
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x12d37e9, 0xe)
	/usr/local/go/src/runtime/panic.go:596 +0x95
runtime.newstack(0x0)
	/usr/local/go/src/runtime/stack.go:1089 +0x3f2
runtime.morestack()
	/usr/local/go/src/runtime/asm_amd64.s:385 +0x86

goroutine 1 [running]:
main.typeOKForCgoCall(0x13efb20, 0xc4203cb920, 0xc420156c30)
	/usr/local/go/src/cmd/vet/cgo.go:116 +0x2d2 fp=0xc440800360 sp=0xc440800358
main.typeOKForCgoCall(0x13efae0, 0xc420156b10, 0xc4203cb950)
	/usr/local/go/src/cmd/vet/cgo.go:129 +0xf4 fp=0xc440800398 sp=0xc440800360
main.typeOKForCgoCall(0x13efb20, 0xc4203cb950, 0xc420156c01)
	/usr/local/go/src/cmd/vet/cgo.go:124 +0x158 fp=0xc4408003d0 sp=0xc440800398

```

**Comment by bradfitz** [comment]
I'll let @ianlancetaylor triage this for Go1.8 vs Go1.9.

**Comment by rfielding** [comment]
The diff is this, but I haven't done the Gerrit setup yet:
```
diff --git a/src/cmd/vet/cgo.go b/src/cmd/vet/cgo.go
index d233e9a..596fef5 100644
--- a/src/cmd/vet/cgo.go
+++ b/src/cmd/vet/cgo.go
@@ -110,26 +110,35 @@ func cgoBaseType(f *File, arg ast.Expr) types.Type {
        return f.pkg.types[arg].Type
 }
 
-// typeOKForCgoCall returns true if the type of arg is OK to pass to a
-// C function using cgo. This is not true for Go types with embedded
-// pointers.
-func typeOKForCgoCall(t types.Type) bool {
+func typeOKForCgoCallRecursive(t types.Type, visited map[types.Type]bool) bool {
        if t == nil {
                return true
        }
+       if visited[t] {
+               return false
+       }
+       visited[t] = true
        switch t := t.Underlying().(type) {
        case *types.Chan, *types.Map, *types.Signature, *types.Slice:
                return false
        case *types.Pointer:
-               return typeOKForCgoCall(t.Elem())
+               return typeOKForCgoCallRecursive(t.Elem(), visited)
        case *types.Array:
-               return typeOKForCgoCall(t.Elem())
+               return typeOKForCgoCallRecursive(t.Elem(), visited)
        case *types.Struct:
                for i := 0; i < t.NumFields(); i++ {
-                       if !typeOKForCgoCall(t.Field(i).Type()) {
+                       if !typeOKForCgoCallRecursive(t.Field(i).Type(), visited) {
                                return false
                        }
                }
        }
        return true
 }
+
+// typeOKForCgoCall returns true if the type of arg is OK to pass to a
+// C function using cgo. This is not true for Go types with embedded
+// pointers.
+func typeOKForCgoCall(t types.Type) bool {
+       visited := make(map[types.Type]bool)
+       return typeOKForCgoCallRecursive(t, visited)
+}
```

**Comment by ianlancetaylor** [comment]
@rfielding Thanks, but please don't send us patches except through Gerrit, since Gerrit ensures that we have the right to use the patch.

**Comment by ianlancetaylor** [comment]
@bradfitz Looks like CL 32754 may have caused cmd/vet to stop running the cgo tests.  Hmmm.

**Comment by ianlancetaylor** [comment]
I won't insist on it, but I think it would be better to fix this for 1.8 since it's fairly annoying to have vet crash on valid code and there is no sensible workaround.
